bakery/store.go

```diff
@@ -0,0 +1,144 @@
+package bakery
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+	"gopkg.in/macaroon.v2-unstable"
+)
+
+// RootKeyStore defines store for macaroon root keys.
+type RootKeyStore interface {
+	// Get returns the root key for the given id.
+	// If the item is not there, it returns ErrNotFound.
+	Get(ctxt context.Context, id []byte) ([]byte, error)
+
+	// RootKey returns the root key to be used for making a new
+	// macaroon, and an id that can be used to look it up later with
+	// the Get method.
+	//
+	// Note that the root keys should remain available for as long
+	// as the macaroons using them are valid.
+	//
+	// Note that there is no need for it to return a new root key
+	// for every call - keys may be reused, although some key
+	// cycling is over time is advisable.
+	RootKey(ctxt context.Context) (rootKey []byte, id []byte, err error)
+}
+
+// NewMemRootKeyStore returns an implementation of
+// Store that generates a single key and always
+// returns that from RootKey. The same id ("0") is always
+// used.
+func NewMemRootKeyStore() RootKeyStore {
+	return new(memRootKeyStore)
+}
+
+type memRootKeyStore struct {
+	mu  sync.Mutex
+	key []byte
+}
+
+// Get implements Store.Get.
+func (s *memRootKeyStore) Get(_ context.Context, id []byte) ([]byte, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if len(id) != 1 || id[0] != '0' || s.key == nil {
+		return nil, ErrNotFound
+	}
+	return s.key, nil
+}
+
+// RootKey implements Store.RootKey by
+//always returning the same root key.
+func (s *memRootKeyStore) RootKey(context.Context) (rootKey, id []byte, err error) {
```

Comment:
Good question. Passing the context in here opens up the possibility of storing root keys in entity documents. The scenario I'm thinking of is this: we make a request to operate on an entity, which causes us to fetch the identity document. When retrieving the root key for the entity, we don't want to make another round trip to the database, so we attach the root key from the entity document to the context. Then RootKey can see that root key in the context and use it without needing to make a request to the database.

Essentially, passing the context in here allows any situation where the root key might be context-dependent.